cli: make --global flag for demo public

otan opened the pull request:
See individual commits for details.

Refs: #62025

otan committed:
cli: make --global flag for demo public

Release note (cli change): The --global flag for cockroach demo is now
advertised. This flag simulates latencies in multi-node demo clusters
when the nodes are set in different regions to simulate real-life global
latencies.

otan committed:
cli: add test for simulated latencies

We previously had no tests simulating the latency in demo using the
--global flag.

Release note: None

otan committed:
cli: disallow adding or shutting down nodes in demo --global

In addition to the release note, reworded the add node (which already
errored beforehand), and adding a clause in RestartNode.

Note there is no code path that allows a restart, since shutdown is
required before a node restart.

Release note (cli change): Previously, `\demo shutdown <node_idx>` would
error if `--global` was set. This will now error gracefully as an
unsupported behavior.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/62435)
<!-- Reviewable:end -->

otan commented:
not sure how to debug that test failure, can't repro it locally or on roachprod :\

otan commented:
ah got one! @knz would you know what this is about:
```
=== CONT  TestTransientClusterSimulateLatencies
    demo_test.go:217: Leaked goroutine: goroutine 6942 [chan receive]:
        net/http.(*persistConn).addTLS(0xc004752a20, 0xc00472d8e0, 0x18, 0x0, 0xc00472d8f9, 0x3)
        	/usr/local/go/src/net/http/transport.go:1515 +0x1a5
        net/http.(*Transport).dialConn(0xc000cec000, 0x53ffbe0, 0xc00453b200, 0x0, 0xc00128b0e0, 0x5, 0xc00472d8e0, 0x1c, 0x0, 0xc004752a20, ...)
        	/usr/local/go/src/net/http/transport.go:1585 +0x1d25
        net/http.(*Transport).dialConnFor(0xc000cec000, 0xc000e75d90)
        	/usr/local/go/src/net/http/transport.go:1421 +0xc6
        created by net/http.(*Transport).queueForDial
        	/usr/local/go/src/net/http/transport.go:1390 +0x40f
        Leaked goroutine: goroutine 7106 [IO wait]:
        internal/poll.runtime_pollWait(0x7f2cf9c6cc28, 0x72, 0x5372300)
        	/usr/local/go/src/runtime/netpoll.go:222 +0x55
        internal/poll.(*pollDesc).wait(0xc009b6a098, 0x72, 0x5372300, 0x7488f98, 0x0)
        	/usr/local/go/src/internal/poll/fd_poll_runtime.go:87 +0x45
        internal/poll.(*pollDesc).waitRead(...)
        	/usr/local/go/src/internal/poll/fd_poll_runtime.go:92
        internal/poll.(*FD).Read(0xc009b6a080, 0xc00178d680, 0x205, 0x205, 0x0, 0x0, 0x0)
        	/usr/local/go/src/internal/poll/fd_unix.go:159 +0x1a5
        net.(*netFD).Read(0xc009b6a080, 0xc00178d680, 0x205, 0x205, 0x203002, 0x58, 0xc000a00000)
        	/usr/local/go/src/net/fd_posix.go:55 +0x4f
        net.(*conn).Read(0xc00ca528a8, 0xc00178d680, 0x205, 0x205, 0x0, 0x0, 0x0)
        	/usr/local/go/src/net/net.go:182 +0x8e
        crypto/tls.(*atLeastReader).Read(0xc009426b80, 0xc00178d680, 0x205, 0x205, 0xc00178d680, 0x0, 0xc001574688)
        	/usr/local/go/src/crypto/tls/conn.go:779 +0x62
        bytes.(*Buffer).ReadFrom(0xc006e7b080, 0x53660c0, 0xc009426b80, 0x43c1a5, 0x40ba480, 0x45dde60)
        	/usr/local/go/src/bytes/buffer.go:204 +0xb1
        crypto/tls.(*Conn).readFromUntil(0xc006e7ae00, 0x536e880, 0xc00ca528a8, 0x5, 0xc00ca528a8, 0x100000451d360)
        	/usr/local/go/src/crypto/tls/conn.go:801 +0xf3
        crypto/tls.(*Conn).readRecordOrCCS(0xc006e7ae00, 0xc003a10000, 0x114, 0x120)
        	/usr/local/go/src/crypto/tls/conn.go:608 +0x115
        crypto/tls.(*Conn).readRecord(...)
        	/usr/local/go/src/crypto/tls/conn.go:576
        crypto/tls.(*Conn).readHandshake(0xc006e7ae00, 0xc008a7f216, 0xc008a7f200, 0x10f, 0x180)
        	/usr/local/go/src/crypto/tls/conn.go:992 +0x6d
        crypto/tls.(*Conn).clientHandshake(0xc006e7ae00, 0x0, 0x0)
        	/usr/local/go/src/crypto/tls/handshake_client.go:170 +0x2a7
        crypto/tls.(*Conn).Handshake(0xc006e7ae00, 0x0, 0x0)
        	/usr/local/go/src/crypto/tls/conn.go:1362 +0xc9
        net/http.(*persistConn).addTLS.func2(0x0, 0xc006e7ae00, 0x0, 0xc000a791a0)
        	/usr/local/go/src/net/http/transport.go:1509 +0x45
        created by net/http.(*persistConn).addTLS
        	/usr/local/go/src/net/http/transport.go:1505 +0x177
```

otan commented:
here are the goroutines:
```
goroutines:

goroutine 213 [running]:
github.com/cockroachdb/cockroach/pkg/util/leaktest.AfterTest.func1()
	/go/src/github.com/cockroachdb/cockroach/pkg/util/leaktest/leaktest.go:141 +0x1e9
github.com/cockroachdb/cockroach/pkg/cli.TestTransientClusterSimulateLatencies(0xc000701e00)
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/demo_test.go:211 +0x662
testing.tRunner(0xc000701e00, 0x4daae18)
	/usr/local/go/src/testing/testing.go:1123 +0xef
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:1168 +0x2b3

goroutine 1 [chan receive]:
testing.(*T).Run(0xc000701e00, 0x474bbc3, 0x25, 0x4daae18, 0x4c9e01)
	/usr/local/go/src/testing/testing.go:1169 +0x2da
testing.runTests.func1(0xc000701c80)
	/usr/local/go/src/testing/testing.go:1439 +0x78
testing.tRunner(0xc000701c80, 0xc001827d58)
	/usr/local/go/src/testing/testing.go:1123 +0xef
testing.runTests(0xc001292e40, 0x74854e0, 0x43, 0x43, 0xc00f1f2f5679042b, 0xe1697bfcc, 0x77afbc0, 0x203000)
	/usr/local/go/src/testing/testing.go:1437 +0x2fe
testing.(*M).Run(0xc001182180, 0x0)
	/usr/local/go/src/testing/testing.go:1345 +0x1eb
github.com/cockroachdb/cockroach/pkg/cli_test.TestMain(0xc001182180)
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/main_test.go:34 +0x8e
main.main()
	_testmain.go:225 +0x165

goroutine 8 [chan receive]:
github.com/cockroachdb/cockroach/pkg/util/log.flushDaemon()
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/log_flush.go:75 +0x73
created by github.com/cockroachdb/cockroach/pkg/util/log.init.5
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/log_flush.go:41 +0x35

goroutine 9 [chan receive]:
github.com/cockroachdb/cockroach/pkg/util/log.signalFlusher()
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/log_flush.go:98 +0x12c
created by github.com/cockroachdb/cockroach/pkg/util/log.init.5
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/log_flush.go:42 +0x4d

goroutine 11 [syscall]:
os/signal.signal_recv(0x0)
	/usr/local/go/src/runtime/sigqueue.go:147 +0x9d
os/signal.loop()
	/usr/local/go/src/os/signal/signal_unix.go:23 +0x25
created by os/signal.Notify.func1.1
	/usr/local/go/src/os/signal/signal.go:150 +0x45

goroutine 21 [select]:
go.opencensus.io/stats/view.(*worker).start(0xc000b823c0)
	/go/src/github.com/cockroachdb/cockroach/vendor/go.opencensus.io/stats/view/worker.go:154 +0x105
created by go.opencensus.io/stats/view.init.0
	/go/src/github.com/cockroachdb/cockroach/vendor/go.opencensus.io/stats/view/worker.go:32 +0x57

goroutine 9877 [IO wait]:
internal/poll.runtime_pollWait(0x7fab07b72c18, 0x72, 0x537e520)
	/usr/local/go/src/runtime/netpoll.go:222 +0x55
internal/poll.(*pollDesc).wait(0xc00b0a5f98, 0x72, 0x537e500, 0x7499f98, 0x0)
	/usr/local/go/src/internal/poll/fd_poll_runtime.go:87 +0x45
internal/poll.(*pollDesc).waitRead(...)
	/usr/local/go/src/internal/poll/fd_poll_runtime.go:92
internal/poll.(*FD).Read(0xc00b0a5f80, 0xc007db0480, 0x205, 0x205, 0x0, 0x0, 0x0)
	/usr/local/go/src/internal/poll/fd_unix.go:159 +0x1a5
net.(*netFD).Read(0xc00b0a5f80, 0xc007db0480, 0x205, 0x205, 0x203003, 0x58, 0xc00078a000)
	/usr/local/go/src/net/fd_posix.go:55 +0x4f
net.(*conn).Read(0xc00b624498, 0xc007db0480, 0x205, 0x205, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/net.go:182 +0x8e
crypto/tls.(*atLeastReader).Read(0xc00d45c120, 0xc007db0480, 0x205, 0x205, 0xc007db0480, 0x0, 0xc00e6b4688)
	/usr/local/go/src/crypto/tls/conn.go:779 +0x62
bytes.(*Buffer).ReadFrom(0xc00d443780, 0x53722c0, 0xc00d45c120, 0x43c1a5, 0x40c3860, 0x45e7a20)
	/usr/local/go/src/bytes/buffer.go:204 +0xb1
crypto/tls.(*Conn).readFromUntil(0xc00d443500, 0x537aaa0, 0xc00b624498, 0x5, 0xc00b624498, 0x1000004526ec0)
	/usr/local/go/src/crypto/tls/conn.go:801 +0xf3
crypto/tls.(*Conn).readRecordOrCCS(0xc00d443500, 0xc0075ab800, 0x114, 0x120)
	/usr/local/go/src/crypto/tls/conn.go:608 +0x115
crypto/tls.(*Conn).readRecord(...)
	/usr/local/go/src/crypto/tls/conn.go:576
crypto/tls.(*Conn).readHandshake(0xc00d443500, 0xc000e9d016, 0xc000e9d080, 0x10f, 0x180)
	/usr/local/go/src/crypto/tls/conn.go:992 +0x6d
crypto/tls.(*Conn).clientHandshake(0xc00d443500, 0x0, 0x0)
	/usr/local/go/src/crypto/tls/handshake_client.go:170 +0x2a7
crypto/tls.(*Conn).Handshake(0xc00d443500, 0x0, 0x0)
	/usr/local/go/src/crypto/tls/conn.go:1362 +0xc9
net/http.(*persistConn).addTLS.func2(0x0, 0xc00d443500, 0x0, 0xc007000e40)
	/usr/local/go/src/net/http/transport.go:1509 +0x45
created by net/http.(*persistConn).addTLS
	/usr/local/go/src/net/http/transport.go:1505 +0x177

goroutine 9927 [chan receive]:
net/http.(*persistConn).addTLS(0xc00d2c3b00, 0xc00d2cf280, 0x18, 0x0, 0xc00d2cf299, 0x3)
	/usr/local/go/src/net/http/transport.go:1515 +0x1a5
net/http.(*Transport).dialConn(0xc000d7c640, 0x540c040, 0xc000923200, 0x0, 0xc008f7f200, 0x5, 0xc00d2cf280, 0x1c, 0x0, 0xc00d2c3b00, ...)
	/usr/local/go/src/net/http/transport.go:1585 +0x1d25
net/http.(*Transport).dialConnFor(0xc000d7c640, 0xc002957340)
	/usr/local/go/src/net/http/transport.go:1421 +0xc6
created by net/http.(*Transport).queueForDial
	/usr/local/go/src/net/http/transport.go:1390 +0x40f
=== CONT  TestTransientClusterSimulateLatencies
    demo_test.go:211: Leaked goroutine: goroutine 9877 [IO wait]:
        internal/poll.runtime_pollWait(0x7fab07b72c18, 0x72, 0x537e520)
        	/usr/local/go/src/runtime/netpoll.go:222 +0x55
        internal/poll.(*pollDesc).wait(0xc00b0a5f98, 0x72, 0x537e500, 0x7499f98, 0x0)
        	/usr/local/go/src/internal/poll/fd_poll_runtime.go:87 +0x45
        internal/poll.(*pollDesc).waitRead(...)
        	/usr/local/go/src/internal/poll/fd_poll_runtime.go:92
        internal/poll.(*FD).Read(0xc00b0a5f80, 0xc007db0480, 0x205, 0x205, 0x0, 0x0, 0x0)
        	/usr/local/go/src/internal/poll/fd_unix.go:159 +0x1a5
        net.(*netFD).Read(0xc00b0a5f80, 0xc007db0480, 0x205, 0x205, 0x203003, 0x58, 0xc00078a000)
        	/usr/local/go/src/net/fd_posix.go:55 +0x4f
        net.(*conn).Read(0xc00b624498, 0xc007db0480, 0x205, 0x205, 0x0, 0x0, 0x0)
        	/usr/local/go/src/net/net.go:182 +0x8e
        crypto/tls.(*atLeastReader).Read(0xc00d45c120, 0xc007db0480, 0x205, 0x205, 0xc007db0480, 0x0, 0xc00e6b4688)
        	/usr/local/go/src/crypto/tls/conn.go:779 +0x62
        bytes.(*Buffer).ReadFrom(0xc00d443780, 0x53722c0, 0xc00d45c120, 0x43c1a5, 0x40c3860, 0x45e7a20)
        	/usr/local/go/src/bytes/buffer.go:204 +0xb1
        crypto/tls.(*Conn).readFromUntil(0xc00d443500, 0x537aaa0, 0xc00b624498, 0x5, 0xc00b624498, 0x1000004526ec0)
        	/usr/local/go/src/crypto/tls/conn.go:801 +0xf3
        crypto/tls.(*Conn).readRecordOrCCS(0xc00d443500, 0xc0075ab800, 0x114, 0x120)
        	/usr/local/go/src/crypto/tls/conn.go:608 +0x115
        crypto/tls.(*Conn).readRecord(...)
        	/usr/local/go/src/crypto/tls/conn.go:576
        crypto/tls.(*Conn).readHandshake(0xc00d443500, 0xc000e9d016, 0xc000e9d080, 0x10f, 0x180)
        	/usr/local/go/src/crypto/tls/conn.go:992 +0x6d
        crypto/tls.(*Conn).clientHandshake(0xc00d443500, 0x0, 0x0)
        	/usr/local/go/src/crypto/tls/handshake_client.go:170 +0x2a7
        crypto/tls.(*Conn).Handshake(0xc00d443500, 0x0, 0x0)
        	/usr/local/go/src/crypto/tls/conn.go:1362 +0xc9
        net/http.(*persistConn).addTLS.func2(0x0, 0xc00d443500, 0x0, 0xc007000e40)
        	/usr/local/go/src/net/http/transport.go:1509 +0x45
        created by net/http.(*persistConn).addTLS
        	/usr/local/go/src/net/http/transport.go:1505 +0x177
        Leaked goroutine: goroutine 9927 [chan receive]:
        net/http.(*persistConn).addTLS(0xc00d2c3b00, 0xc00d2cf280, 0x18, 0x0, 0xc00d2cf299, 0x3)
        	/usr/local/go/src/net/http/transport.go:1515 +0x1a5
        net/http.(*Transport).dialConn(0xc000d7c640, 0x540c040, 0xc000923200, 0x0, 0xc008f7f200, 0x5, 0xc00d2cf280, 0x1c, 0x0, 0xc00d2c3b00, ...)
        	/usr/local/go/src/net/http/transport.go:1585 +0x1d25
        net/http.(*Transport).dialConnFor(0xc000d7c640, 0xc002957340)
        	/usr/local/go/src/net/http/transport.go:1421 +0xc6
        created by net/http.(*Transport).queueForDial
        	/usr/local/go/src/net/http/transport.go:1390 +0x40f
--- FAIL: TestTransientClusterSimulateLatencies (13.55s)
    --- PASS: TestTransientClusterSimulateLatencies/from_us-east1 (0.59s)
    --- PASS: TestTransientClusterSimulateLatencies/from_us-west1 (0.41s)
    --- PASS: TestTransientClusterSimulateLatencies/from_europe-west1 (0.53s)
FAIL
```

otan commented:
I wonder if this is because the leak detector takes 5s to detect leaks, but due to the RPCs being slowed down & this being longer, that's not enough time. in which case, we should ignore the leak detector.

@knz do you reckon this is feasible or nah?

i'm still curious why this only reproduces under `make roachprod-stress` but not locally for me....

otan commented:
changing the leaktest timeout to 60s seems to unblock some things, but introduce a panic during the stop phase: https://gist.github.com/otan/12e1e6d7aab9236e3590660ef570aee8

otan commented:
Yeah spent too long on this. Maybe one day...

knz reviewed (commented):
> Reviewed 1 of 1 files at r1, 2 of 2 files at r2, 3 of 3 files at r3, 1 of 1 files at r4, 1 of 1 files at r5, 2 of 4 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62435)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz and @otan)

___
*[pkg/cli/demo_cluster.go, line 129 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/62435#-MWd6DkAEnVqYUhNrHlx:-MWd6DkB7GNKF3RqOI2B:bnpov6v) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab655fa82b8e330e69bc3bcc9e0335ca0f691b1f/pkg/cli/demo_cluster.go#L129)):*
> ```Go
> 	latencyMapWaitCh := make(chan struct{})
> 
> 	// serverReadyCh is used .
> ```

nit: sentence in comment
___
*[pkg/cli/demo_cluster.go, line 190 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/62435#-MWd6zDm8R1PZNmHt1QT:-MWd6zDm8R1PZNmHt1QU:b-ykrx2p) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab655fa82b8e330e69bc3bcc9e0335ca0f691b1f/pkg/cli/demo_cluster.go#L190)):*
> ```Go
> 
> 		// We force a wait for all servers until they are ready.
> 		nodeReadyCh := make(chan struct{}, 1)
> ```

the channel handling throughout this function is super confusing (and probably confused)

I would encourage you to do the following. For each channel:

1. document at the point the channel is created, what are the readers and what are the writers, and the exact conditions under which channels are read and written

2. where it makes sense, use `close(chan)` on the writer to ensure that the number of readers doesn't matter afterwards.
___
*[pkg/cli/demo_cluster.go, line 202 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/62435#-MWd6VNj8Z5Vv779xsmc:-MWd6VNj8Z5Vv779xsmd:b-by6frt) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab655fa82b8e330e69bc3bcc9e0335ca0f691b1f/pkg/cli/demo_cluster.go#L202)):*
> ```Go
> 				err := serv.Start()
> 				nodeErrCh <- err
> 				errCh <- err
> ```

this channel handling is messed up. when err !=nil, the write to `errCh` will block and this goroutine will remain running.
___
*[pkg/cli/demo_cluster.go, line 407 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/62435#-MWd7StY07LEl7gsMVsF:-MWd7StY07LEl7gsMVsG:besscub) ([raw file](https://github.com/cockroachdb/cockroach/blob/83df4586e7467951d1aab9e1d8579757a3beb6ca/pkg/cli/demo_cluster.go#L407)):*
> ```Go
> func (c *transientCluster) DrainAndShutdown(nodeID roachpb.NodeID) error {
> 	if demoCtx.simulateLatency {
> 		return errors.Errorf("shutting down nodes is not supported in --global configurations")
> ```

here and throughout the code: 
1. it's super confusing to have the field in the ctx struct named differently than the flag. I may encourage you to find a way to make them match.

2. please factor the check and the definition of the error message in a single function and call it from the multiple places

3. the proper way to construct an error message is `errors.Errorf("... --%s ...", cliflags.YourFlag.Name)`   (or `errors.Newf`)
___
*[pkg/cli/demo_test.go, line 127 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/62435#-MWd7zx-7GQGQKkV9CjO:-MWd7zx0A3NCQplEE4Ln:b-aangw1) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bb02ba05b48629f3ae45b8a6ad039a414369013/pkg/cli/demo_test.go#L127)):*
> ```Go
> 	defer log.Scope(t).Close(t)
> 
> 	// Set the leak test timeout to be 60s, as due to the delay
> ```

That should not be necessary, if all the close / cleanup calls are performed properly. If you think you need this, that means that a close function is not being called, or a goroutine is left to hang for longer than necessary.


<!-- Sent from Reviewable.io -->

otan commented:
As I'm done with this for now, hints for the next person.

Spent one more day working on tests `demo --global`. I'm done with it - as in - I don't think I can get it in. There are two remaining issues with the bespoke test setup:
* It panics during shutdown as it does not receive the magic bytes header. I don't know where it comes from, but know how to reproduce it quickly (\demo shutdown X during --global). I've disabled `\demo shutdown X` in one of the commits here, but the error still pops up when you try cleanup the cluster.
* Sometimes there is a leaking goroutine from grpc's AddTLS. The goroutine trace doesn't show anything obvious, so I have no idea what is wrong here. Can't find the spawning goroutine (not obvious from the stack traces), and this code is all in a third party library.

otan commented:
___
*[pkg/cli/demo_cluster.go, line 202 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/62435#-MWd6VNj8Z5Vv779xsmc:-MWfBHmWBJfx4cWCnbhv:b53binp) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab655fa82b8e330e69bc3bcc9e0335ca0f691b1f/pkg/cli/demo_cluster.go#L202)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

this channel handling is messed up. when err !=nil, the write to `errCh` will block and this goroutine will remain running.
</blockquote></details>

both channels are buffered, so i don't think we ever block.


<!-- Sent from Reviewable.io -->

knz commented:
Ok thank you for all your work and your investigative steps.
Let's pick this back up when the roadmap planning work has settled a bit.

otan commented:
> It panics during shutdown as it does not receive the magic bytes header.

this is now resolved. just the leaking goroutine left.

knz commented:
There are multiple goroutines leaked, but this is clearly the smoking gun:

```
    demo_test.go:213: Leaked goroutine: goroutine 1111 [select]:
        github.com/cockroachdb/cockroach/pkg/util/retry.(*Retry).Next(0xc001277d38, 0xc0035964e0)
                /data/home/kena/src/go/src/github.com/cockroachdb/cockroach/pkg/util/retry/retry.go:128 +0x151
        github.com/cockroachdb/cockroach/pkg/sqlmigrations.(*Manager).EnsureMigrations(0xc0013a5980, 0x1f8f770, 0xc0035964e0, 0x200000014, 0x3000000000, 0x0, 0x0)
                /data/home/kena/src/go/src/github.com/cockroachdb/cockroach/pkg/sqlmigrations/migrations.go:578 +0x3d6
        github.com/cockroachdb/cockroach/pkg/server.(*SQLServer).preStart(0xc004b03400, 0x1f8f770, 0xc0035964e0, 0xc002a9a600, 0x1f07280, 0xc0014bdba0, 0x0, 0x0, 0x0, 0x0, ...)
                /data/home/kena/src/go/src/github.com/cockroachdb/cockroach/pkg/server/server_sql.go:839 +0x738
        github.com/cockroachdb/cockroach/pkg/server.(*Server).PreStart(0xc00458e800, 0x1f8f700, 0xc000074188, 0x0, 0x0)
                /data/home/kena/src/go/src/github.com/cockroachdb/cockroach/pkg/server/server.go:1822 +0x3408
        github.com/cockroachdb/cockroach/pkg/server.(*Server).Start(0xc00458e800, 0x1f8f700, 0xc000074188, 0x0, 0x0)
                /data/home/kena/src/go/src/github.com/cockroachdb/cockroach/pkg/server/server.go:1108 +0x45
        github.com/cockroachdb/cockroach/pkg/server.(*TestServer).Start(...)
                /data/home/kena/src/go/src/github.com/cockroachdb/cockroach/pkg/server/testserver.go:465
        github.com/cockroachdb/cockroach/pkg/cli.(*transientCluster).start.func2(0xc0042cd680, 0xc004e30e40, 0xc0017b92c0, 0xc002d3b8c0, 0xc00004aae0)
                /data/home/kena/src/go/src/github.com/cockroachdb/cockroach/pkg/cli/demo_cluster.go:203 +0x4f
        created by github.com/cockroachdb/cockroach/pkg/cli.(*transientCluster).start
                /data/home/kena/src/go/src/github.com/cockroachdb/cockroach/pkg/cli/demo_cluster.go:202 +0x67d
```

In other words, there's this goroutine launched under `if demoCtx.simulateLatencies` in `(*transientCluster).start()` which does not properly stop even after the test complete.

I already mentioned that the pattern of interlocking channels was problematic in my previous review.

I will now investigate this further—exactly like I suggested in my previous review: by analyzing exactly where they get written and read, and discover which synchronization point is missing—and report my findings.

knz commented:
@otan can you check "allow edits by maintainer" on this PR, so that I can push to your branch. I think I fixed your code here: https://github.com/knz/cockroach/commits/otan_demo_test

knz commented:
Request by @jordanlewis

knz commented:
For teaching purposes, here is (at a high level) how I approached this troubleshooting session:

- I started by simply observing that the demo cluster initialization was too complex to read and understand. So I first reverse-engineered it.
- then I added a comment at the top of the function that explains the sequence of steps performed, as found by reverse-engineering.
- then I converted the comment into a sequence of function calls, one per step (or sub-step)
- then I added logging calls at each step, so I could monitor the progress of the initialization by looking at logs.
- then I moved the original code relating to each sub-step to its own sub-function.

This was merely a code transformation and I was not doing functional changes as I was doing so (So I expected the test to continue failing).

During this process, I also extended the demo and server code to use the standard context cancellation / stopper shutdown protocol we use throughout the server package: whenever a channel is read from or written to, also check the `ctx.Done` and `stopper.ShouldQuiesce()` alongside, no questions asked.

Then after the code was reworked as per the previous step (I considered this to be merely a general refactor to increase readability and troubleshootability), I started looking at the failing test.

The first observation I made then is that the leaking goroutine problem had disappeared. This is likely thanks to the disciplined use of `stopper.RunAsyncTask` instead of a naked `go` keyword, as well as disciplined reads of the `timeoutCh`, `ctx.Done()` and `stopper.ShouldQuiesce()` channels.

However this revealed a new failure mode, where the test would non-deterministically fail by failing to initialize the `demo` account sometimes.

Thanks to the new structure of the code and the new “phase-by-phase logging”, I discovered that the servers were reporting readiness via the `ReadyFn` earlier than the demo code was assuming (they report readiness when the RPC subsystem is ready, not when SQL is ready). 

Then I switched the readiness check to use the standard health probe instead, and that solved that problem.

otan commented:
I don't have that button. You should be able to force push. Thanks for looking into this!

otan commented:
(if not ill try manually incorporate them in)

knz commented:
The check box is at the bottom right of the PR metadata:
![image](https://user-images.githubusercontent.com/642886/113474121-33204b80-946e-11eb-93a6-b4e62daeb836.png)

otan commented:
![image](https://user-images.githubusercontent.com/3646147/113474732-7b1a8f00-94bd-11eb-8e96-b618ed380ed1.png)
i don't see the button :\

otan commented:
no worries, let me do it and rename the flags!

otan commented:
oh i see you already did that.

otan reviewed (commented):
your changes LGTM @knz, feel free to bors r+!

knz commented:
> feel free to bors r+!

thanks, I'm going to let CI run to completion first.

knz reviewed (commented):
> Reviewed 5 of 14 files at r7, 2 of 4 files at r8, 2 of 3 files at r10, 1 of 2 files at r11, 5 of 5 files at r12.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62435)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)



<!-- Sent from Reviewable.io -->

knz commented:
ok so there are some CI failure, I'll look into it

knz commented:
ok so the full concurrent initialization doesn't work well with node ID assignments, it makes them non-deterministic. I'll work on this to fix the tests in #63057 and then we can rebase this one.

otan commented:
Sounds good. Any idea what the goroutine leak is?

knz commented:
> Any idea what the goroutine leak is? 

huh that looks like a new one. will also look into it

knz commented:
@otan can you rebase this branch on top of #63057? 
then also fix the one linter failures that is remaining in `demo_test.go`

with the rebase I am unable to repro the leaking goroutine on my local machine. Can you stress it locally too? Do you observe the leaking goroutine on your system?

otan commented:
rebased. it looks the same as the leaking goroutine problem i've been having all along. i can only reproduce it using:

```
 roachprod create -n 1 $CLUSTER --gce-machine-type n1-standard-16   --lifetime 4h --gce-zones australia-southeast1-a
make roachprod-stress PKG=./pkg/cli TESTS=TestTransientClusterGlobal CLUSTER=$CLUSTER TESTTIMEOUT=1m TESTFLAGS='--show-logs -v' |& tee ~/out
```

otan commented:
(off to sleep for several hours! thanks for looking into this)

knz commented:
Cool I hadn't thought of using roach prod. I'll try that too.

knz commented:
Some notes as I am investigating this:

- I am unable to repro while stressing with go 1.16.2 (~1000 runs).
- FYI with your original code, the leaked goroutine that you reported does not happen with go 1.16, however another one (which is fixed by #63057) happens immediately (~20 runs). All tests now are running with the latest code on this branch.
- I have noticed that there is a *panic* with a similar stack trace when using a local go (1.15.8) after ~60 runs. (See below) 
- you reported the failure using `roachprod-stress`. This uses the version of go 1.15 in the builder image   (1.15.6). So I tried with `build/builder.sh make stress ...` first. This does not repro the goroutine leak, but does repro the panic (~300 runs).
- `roachprod-stress` builds the test binary in "release" mode (whereas `make stress` uses dev mode). So I did this, and than ran the test manually: `bin.docker_amd64/stress ./cli.test -test.run "TestTransientClusterSim" -test.v -test.timeout 1m` - then I got the panic again, same stack trace, also after ~60 runs (but no leaked goroutine).
- then I ran the same command `./stress ./cli.test ...` on the roachprod machine, and I observed the same leaked goroutine as you after ~200 runs.

Here's the summary so far:

| Test | computer | Go version | Result |
|---|---|---|---|
| `make stress ...` | local | 1.16.2 | test passes |
| `make stress ...` | local | 1.15.8 | panic |
| `build/builder.sh make stress ...` | local | 1.15.6 (builder image) | panic |
| using release mode binaries, `stress ./cli.test ...` | local | 1.15.6 (builder image) | panic |
| `make roachprod-stress ...` | roachprod server | 1.15.6 (builder image) | goroutine leak |

Panic message:

```
=== RUN   TestTransientClusterSimulateLatencies/from_europe-west1
E210405 09:04:19.983639 10906 kv/kvserver/queue.go:1093  [n1,replicate,s1,r9/1:/Table/1{3-4}] 1  change replicas of r9 failed: fetching current range descriptor value: node unavailable; try another peer
panic: didn't get expected magic bytes header

goroutine 11367 [running]:
github.com/cockroachdb/cockroach/pkg/rpc.(*delayingConn).Read(0xc000fab440, 0xc001626fc0, 0x205, 0x205, 0x0, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/rpc/context.go:913 +0x393
crypto/tls.(*atLeastReader).Read(0xc001037e60, 0xc001626fc0, 0x205, 0x205, 0xc001626fc0, 0x0, 0xc00e3346a8)
        /usr/local/go/src/crypto/tls/conn.go:779 +0x62
bytes.(*Buffer).ReadFrom(0xc009118600, 0x5398200, 0xc001037e60, 0x43c1a5, 0x40e0d60, 0x43bd6c0)
        /usr/local/go/src/bytes/buffer.go:204 +0xb1
crypto/tls.(*Conn).readFromUntil(0xc009118380, 0x7fcdce218250, 0xc000fab440, 0x5, 0xc000fab440, 0x4726dc)
        /usr/local/go/src/crypto/tls/conn.go:801 +0xf3
crypto/tls.(*Conn).readRecordOrCCS(0xc009118380, 0xc000c11300, 0xf4, 0x100)
        /usr/local/go/src/crypto/tls/conn.go:608 +0x115
crypto/tls.(*Conn).readRecord(...)
        /usr/local/go/src/crypto/tls/conn.go:576
crypto/tls.(*Conn).readHandshake(0xc009118380, 0xc000d87616, 0xc000d87680, 0xef, 0x180)
        /usr/local/go/src/crypto/tls/conn.go:992 +0x6d
crypto/tls.(*Conn).clientHandshake(0xc009118380, 0x0, 0x0)
        /usr/local/go/src/crypto/tls/handshake_client.go:170 +0x2a7
crypto/tls.(*Conn).Handshake(0xc009118380, 0x0, 0x0)
        /usr/local/go/src/crypto/tls/conn.go:1362 +0xc9
google.golang.org/grpc/credentials.(*tlsCreds).ClientHandshake.func1(0xc003a73bc0, 0xc009118380)
        /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/credentials/tls.go:88 +0x2b
created by google.golang.org/grpc/credentials.(*tlsCreds).ClientHandshake
        /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/credentials/tls.go:87 +0xe5


ERROR: exit status 2
```

knz commented:
I looked at this panic first. The panic is in our simulated latency code. This is a bug: the code does not properly handle the case when the RPC connection is interrupted by the remote server shutting down prematurely. (Which happens in this test.) So we need to correct that first. I'll make a commit for that then re-run the tests.

otan commented:
Interesting. The build images should be 1.15.10 now: https://github.com/cockroachdb/cockroach/blob/6e5dfc46eab59f7fd6e77ee5dbab5bac715330de/build/builder/Dockerfile#L238

The [release notes for Go](https://golang.org/doc/devel/release.html#go1.15) indicate numerous bug fixes for net/http, which this is stuck on. I wonder if it's related.

knz commented:
With the panic issue fixed (#63076), I get the following result:

| Test | computer | Go version | Result |
|---|---|---|---|
| `make stress ...` | local | 1.16.2 | test passes |
| `make stress ...` | local | 1.15.8 | test passes |
| `build/builder.sh make stress ...` | local | 1.15.6 (builder image) | test passes |
| using release mode binaries, `stress ./cli.test ...` | local | 1.15.6 (builder image) | test passes |

Not tested via roachprod yet.

> The build images should be 1.15.10 now

Not on your branch, which hasn't been rebased in a while :)

I'll do a rebase on top of #63076 (which is itself rebased on top of `master`) then report results.

knz commented:
Ok so before the rebase, with just the fix in #63076, the goroutine leak on the roachprod machine still happens. I'll now do the rebase and try again.

knz commented:
After rebase:

| Test | computer | Go version | Result |
|---|---|---|---|
| `build/builder.sh make stress ...` | local | 1.15.10 (builder image) | test passes |
| using release mode binaries, `stress ./cli.test ...` | local | 1.15.10 (builder image) | goroutine leak |

ok so at last with the latest builder image I am observing the goroutine leak locally too! This is progress.

knz commented:
@otan, a quesiton completely unrelated to the issue at hand:

```go
      // Find the maximum latency in the cluster from the current node.
      var maxLatency time.Duration
      for _, latencyMS := range regionToRegionToLatency["us-east1"] {
        if d := time.Duration(latencyMS) * time.Millisecond; d > maxLatency {
          maxLatency = d
        }
      }
```

What is this code supposed to do? The comment refers to "from the current node" but the loop does not take the current node into account?

knz commented:
I also think it should be `time.Duration(latencyMS * time.Millisecond)` not `time.Duration(latencyMS) * time.Millisecond`

otan commented:
> What is this code supposed to do? The comment refers to "from the current node" but the loop does not take the current node into account?

Ah woops! This was meant to be from the current node, but when I first wrote the test I only did it from node 0! I'll fix that up.

otan commented:
```
diff --git a/pkg/cli/demo_test.go b/pkg/cli/demo_test.go
index a8fdb46f8d..42e07ea279 100644
--- a/pkg/cli/demo_test.go
+++ b/pkg/cli/demo_test.go
@@ -195,7 +195,7 @@ func TestTransientClusterGlobal(t *testing.T) {
                        defer conn.Close()
                        // Find the maximum latency in the cluster from the current node.
                        var maxLatency time.Duration
-                       for _, latencyMS := range regionToRegionToLatency["us-east1"] {
+                       for _, latencyMS := range regionToRegionToLatency[tc.region] {
                                if d := time.Duration(latencyMS) * time.Millisecond; d > maxLatency {
                                        maxLatency = d
                                }
```

if that helps! i'll rebase after your stuff merges

knz commented:
Here's a new batch of results:

After rebase on top of #63082:

| Test | computer | Go version | Result |
|---|---|---|---|
| `build/builder.sh make stress ...` | local | 1.15.10 (builder image) | test passes (~9000 runs) |
| using release mode binaries, `stress ./cli.test ...` | local | 1.15.10 (builder image) | test passes (~1000 runs) |

So I am moderately confident that #63082 fixed the leak. 

Let's see what CI on #63033 thinks about it.

knz commented:
(Meanwhile the CI on #63081 (the refactor) which is included in my new batch of results is failing due to some regression introduced by #63081. I will rework #63082 without that refactor.)

knz commented:
ok so I'm very sad because the fix in #63082 is sufficient for roachprod, local stress etc
but as shown in #63033 (https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_UnitTests/2848483?buildTab=overview) the test still fails in CI.

I'll try something else.

knz commented:
ok so we're back to square one: this is a really a stuck goroutine, not just a transient goroutine that's about to disappear.

I found this out by looking at the leaktest code. The checker already waits for 5 seconds for stray goroutines to disappear before reporting a leak. This is much larger than the simulated latencies.

So if it finds something, a goroutine has been around for 5 seconds or more *after the test completed*. That's much larger than the simulated latencies so a small delay will not help.

knz commented:
What's remarkable is that we now have 6 different environments (see table above) where the test passes under stress, and it's just in CI that it doesn't.

knz commented:
@otan there's a more fundamental problem going on here, one that suggests this work is relying on shaky foundations: #63098   

I would recommend to first create some solid unit testing for `delayingConn` in `rpc/context_test.go`, which exercises this code for both secure and insecure connections; and then once that works (and is proven not to leak goroutines), continue with the PR here.

jordanlewis commented:
@knz --insecure is deprecated in demo, so I don't understand why you are insisting we should try to fix the global latencies for it, since global latencies is a feature only for demo.

jordanlewis commented:
Oops, I will re-create this comment on the issue you created.

knz commented:
See this comment: https://github.com/cockroachdb/cockroach/pull/63033#issuecomment-824860848

We can now make progress on this, thanks to #63853 being merged.

otan commented:
thanks for the heads up!

do we still want to fix SHOW ALL CLUSTER QUERIES in --insecure before making it public? FWIW, from experimentation it's just that one that doesn't respect latency but i'm not sure why :\

knz commented:
> do we still want to fix SHOW ALL CLUSTER QUERIES in --insecure before making it public?

I think we can make `--global` public but it still needs to be marked as experimental (at least in the docstring that shows up in `--help` and in docs)

otan commented:
ok, i've pulled your commit and added one on top @knz!

knz reviewed (commented):
Ok so this is now on the finish line.

Let's discuss some specifics:

- I'm not too fond of the last commit. Even if we accept the command-line flag's name for the sake of PR and marketing, we still want our code to be descriptive. A boolean field should indicate what the boolean is about. I don't believe that the word "global" is descriptive of anything. In comparison, "simulate latencies" is better. I'd say "simulateGeoLatencies" would even be better.

- I'm curious if there is a way to announce the fact the cluster is not "local" when the session starts. Maybe via the prompt? or a banner when the shell starts?

> Reviewed 15 of 15 files at r16.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62435#-:-MZIDkuc7dNn8QWWUXAR:b-5b08qt)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz and @otan)



<!-- Sent from Reviewable.io -->

otan commented:
i'd be happy to rename it if we also changed the flag as appropriate, but that would also ideally involve some PM input. (the original complaint i was addressing was the flag name and variable name did not match)
i can also just take out the last commit for now and we can table that for another time.

knz commented:
> the original complaint i was addressing was the flag name and variable name did not match

I may have missed that - where was this complaint voiced? (my apologies if I was the person who voiced it, I likely have forgotten)

We have this situation in a couple of other places, and I agree we probably want to do something about it, but maybe we can do this in a principled manner all at once.

>  i can also just take out the last commit for now and we can table that for another time.

if there's an issue filed for it, yes that would make sense.

otan committed:
cli: add startup text on cockroach demo --global

Release note (cli change): There will now be a message upon start-up on
cockroach demo --global indicating latencies between nodes will simulate
real world latencies.

otan commented:
> I may have missed that - where was this complaint voiced? (my apologies if I was the person who voiced it, I likely have forgotten)

would be from https://github.com/cockroachdb/cockroach/pull/62435#pullrequestreview-620999871

here and throughout the code:

it's super confusing to have the field in the ctx struct named differently than the flag. I may encourage you to find a way to make them match.

> if there's an issue filed for it, yes that would make sense.

alrighty. #64269, removed the commit.

> I'm curious if there is a way to announce the fact the cluster is not "local" when the session starts. Maybe via the prompt? or a banner when the shell starts?

I've added some help text.

knz reviewed (approved):
> Reviewed 5 of 5 files at r17.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62435)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz and @otan)



<!-- Sent from Reviewable.io -->

otan commented:
thanks for the guidance!

bors r=knz
